src/librustc_typeck/astconv.rs

```diff
@@ -337,112 +341,128 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+                err.emit();
 
-            let mut err = tcx.sess.struct_span_err_with_code(
-                spans.clone(),
-                &format!(
-                    "wrong number of {} arguments: expected {}{}, found {}",
-                    kind, quantifier, bound, provided,
-                ),
-                DiagnosticId::Error("E0107".into()),
-            );
-            for span in spans {
-                err.span_label(span, label.as_str());
-            }
-            err.emit();
+                true
+            };
 
-            (
-                provided > required, // `suppress_error`
-                potential_assoc_types,
-            )
-        };
+        let mut arg_count_mismatch = reported_late_bound_region_err;
+        let mut unexpected_spans = vec![];
 
-        if reported_late_bound_region_err.is_none()
+        if !reported_late_bound_region_err
             && (!infer_lifetimes || arg_counts.lifetimes > param_counts.lifetimes)
         {
-            check_kind_count(
+            arg_count_mismatch |= check_kind_count(
                 "lifetime",
                 param_counts.lifetimes,
                 param_counts.lifetimes,
                 arg_counts.lifetimes,
                 0,
+                &mut unexpected_spans,
             );
         }
         // FIXME(const_generics:defaults)
         if !infer_args || arg_counts.consts > param_counts.consts {
-            check_kind_count(
+            arg_count_mismatch |= check_kind_count(
                 "const",
                 param_counts.consts,
                 param_counts.consts,
                 arg_counts.consts,
                 arg_counts.lifetimes + arg_counts.types,
+                &mut unexpected_spans,
             );
         }
         // Note that type errors are currently be emitted *after* const errors.
         if !infer_args || arg_counts.types > param_counts.types - defaults.types - has_self as usize
         {
-            check_kind_count(
+            arg_count_mismatch |= check_kind_count(
                 "type",
                 param_counts.types - defaults.types - has_self as usize,
                 param_counts.types - has_self as usize,
                 arg_counts.types,
                 arg_counts.lifetimes,
-            )
-        } else {
-            (reported_late_bound_region_err.unwrap_or(false), None)
+                &mut unexpected_spans,
+            );
         }
+
+        (if arg_count_mismatch { Err(GenericArgCountMismatch) } else { Ok(()) }, unexpected_spans)
+    }
+
+    /// Report an error that a generic argument did not match the generic parameter that was
+    /// expected.
+    fn generic_arg_mismatch_err(sess: &Session, arg: &GenericArg<'_>, kind: &'static str) {
+        struct_span_err!(
+            sess,
+            arg.span(),
+            E0747,
+            "{} provided when a {} was expected",
+            arg.descr(),
+            kind,
+        )
+        .emit();
```

Comment:
No reordering suggestion for this case?

Reply:
I'd rather leave this PR for the bug-fixes and follow up with the diagnostics improvements. That's a good follow up, though.